File: Java/OPSJLib/src/ops/TcpServerSender.java
/**
*
* Copyright (C) 2006-2010 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops;

import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.concurrent.Semaphore;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Anton Gravestam
 */
public class TcpServerSender implements Sender, Runnable
{
    private final String serverIp;
    private final int serverPort;
    private final int sendBufferSize;
    private ServerSocket server = null;
    private TcpSenderList tcpSenderList;
    private volatile boolean listening = false;
    private Semaphore sem = new Semaphore(0);

    public TcpServerSender(String serverIp, int serverPort, int sendBufferSize) throws IOException
    {
        this.serverIp = serverIp;
        this.serverPort = serverPort;
        this.sendBufferSize = sendBufferSize;
//moved below        server = new ServerSocket(serverPort, 0, InetAddress.getByName(serverIp));
        tcpSenderList = new TcpSenderList();
        new Thread(this).start();
        open();
    }

    public final void open()
    {
        if (!listening)
        {
            listening = true;       // Set flag before we signal the thread
            sem.release();          // Signal semaphore so the thread is released
        }
    }

    public void close()
    {
        if (listening)
        {
            sem.drainPermits();     // Clear semaphore so thread will hang on it
            listening = false;      // clear flag before we call stop()

            try {
                // Stop listening. The accept() call in Run() will exit
                // and the thread will wait on the event again
                if (server != null) server.close();
            } catch (IOException ex) {
                Logger.getLogger(TcpServerSender.class.getName()).log(Level.SEVERE, null, ex);
            }

            // Clear all connections (sockets) to subscribers
            tcpSenderList.emptyList();
        }
    }

    public boolean sendTo(byte[] bytes, String ip, int port)
    {
        return sendTo(bytes, 0, bytes.length, ip, port);
    }

    public boolean sendTo(byte[] bytes, int offset, int size, String ip, int port)
    {
        try
        {
            return sendTo(bytes, offset, size, InetAddress.getByName(ip), port);
        } catch (UnknownHostException ex)
        {
            Logger.getLogger(TcpServerSender.class.getName()).log(Level.SEVERE, null, ex);
            return false;
        }
    }

    public boolean sendTo(byte[] bytes, int offset, int size, InetAddress ipAddress, int port)
    {
        return tcpSenderList.sendTo(bytes, offset, size, ipAddress, port);
    }

    public void run()
    {
        while(true)
        {
            try {
                /// Wait for semaphore to be set
                sem.acquire();
                
                // this.tcpListener.Start();
                server = new ServerSocket(this.serverPort, 0, InetAddress.getByName(this.serverIp));
                
                while (listening) {
                    try {
                        Socket socket = server.accept();
                        if (sendBufferSize > 0) socket.setSendBufferSize(sendBufferSize);
                        tcpSenderList.add(socket);
                    } catch (IOException ex) {
                        if (listening) Logger.getLogger(TcpServerSender.class.getName()).log(Level.SEVERE, null, ex);
                    }
                }
                server.close();
            } catch (IOException ex) {
                Logger.getLogger(TcpServerSender.class.getName()).log(Level.SEVERE, null, ex);
            } catch (InterruptedException ex) {
                Logger.getLogger(TcpServerSender.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }

}

File: trunk/C#/Ops/TcpServerSender.cs
///////////////////////////////////////////////////////////
//  TcpServerSender.cs
//  Implementation of the Class TcpServerSender
//  Created on:      12-nov-2011 09:25:37
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.Net;
using System.Net.Sockets;

namespace Ops 
{
	public class TcpServerSender : EasyThread, ISender
    {
        private TcpListener tcpListener;
		private readonly string serverIp;
		private readonly int serverPort;
        private readonly int sendBufferSize;
		private TcpSenderList tcpSenderList;
        private bool listening = false;
        private System.Threading.ManualResetEvent signal = new System.Threading.ManualResetEvent(false);

        public TcpServerSender(string serverIp, int serverPort, int sendBufferSize) 
        {
            this.serverIp = serverIp;
            this.serverPort = serverPort;
            this.sendBufferSize = sendBufferSize;

            IPAddress serverAddr = IPAddress.Parse(serverIp);
            this.tcpListener = new TcpListener(serverAddr, serverPort);
            tcpSenderList = new TcpSenderList();
            this.Start();
            Open();
        }

        public void Open()
        {
            if (!listening)
            {
                listening = true;       // Set flag before we signal the thread
                signal.Set();           // Set event so the thread is released
            }
        }

        public void Close()
        {
            if (listening)
            {
                signal.Reset();         // Clear event so thread will hang on it
                listening = false;      // Clear flag before we call stop()

                // Stop listening. The AcceptTcpClient() call in Run() will exit
                // and the thread will wait on the event again
                this.tcpListener.Stop();

                // Clear all connections (sockets) to subscribers
                this.tcpSenderList.EmptyList();
            }
        }

        /// <summary>
        /// Get IP and port used as endpoint when sending a message
        /// </summary>
        /// <param name="IP"></param>
        /// <param name="port"></param>
        public void GetLocalEndpoint(ref string IP, ref int port)
        {
            IP = this.serverIp;
            port = this.serverPort;
        }

        public bool SendTo(byte[] bytes, string ip, int port)
        {
            return SendTo(bytes, 0, bytes.Length, ip, port);
        }

        public bool SendTo(byte[] bytes, int offset, int size, string ip, int port)
        {
            try
            {
                return SendTo(bytes, offset, size, InetAddress.GetByName(ip), port);
            } catch (Exception ex)
            {
                Logger.ExceptionLogger.LogException(ex);
                return false;
            }
        }

        public bool SendTo(byte[] bytes, int offset, int size, InetAddress ipAddress, int port)
        {
            return tcpSenderList.SendTo(bytes, offset, size, ipAddress, port);
        }

        protected override void Run()
        {
            while (true)
            {
                try
                {
                    /// Wait for event to be set
                    this.signal.WaitOne();

                    this.tcpListener.Start();

                    while (listening)
                    {
                        try
                        {
                            // Blocks until a client has connected to the server
                            System.Net.Sockets.TcpClient client = this.tcpListener.AcceptTcpClient();
                            if (this.sendBufferSize > 0) client.SendBufferSize = this.sendBufferSize;
                            tcpSenderList.Add(client);
                        }
                        catch (Exception ex)
                        {
                            if (listening) Logger.ExceptionLogger.LogException(ex);
                        }
                    }
                }
                catch (SocketException ex)
                {
                    signal.Reset();         // Clear event so thread will hang on it
                    if (ex.SocketErrorCode == System.Net.Sockets.SocketError.AddressAlreadyInUse)
                    {
                        Logger.ExceptionLogger.LogMessage("TcpServer [" + this.serverIp + "::" + this.serverPort + "] Error: " + ex.SocketErrorCode.ToString());
                    }
                    else
                    {
                        Logger.ExceptionLogger.LogException(ex);
                    }
                }
            }
        }

	}

}